### PR TITLE
Adding Semantics Traversal Order example.

### PR DESCRIPTION
**Pull request by gspencergoog**
This adds a traversal order example. This can't be landed until PR #4540 is landed.

The resulting app looks like this:
screenshot_20180118-155536

### CODE REVIEW

File: examples/catalog/lib/custom_a11y_traversal.dart [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+/// they can also override those of their parents (by setting
+/// [SemanticsSortOrder.discardParentOrder] to true), and an entire sort order can be
+/// defined with multiple keys, to provide for virtually any ordering.
+///
+/// Keys at the same position in the sort order are compared with each other, and
+/// keys which are of different types, or which have different [SemanticSortKey.name]
+/// values compare as "equal" so that two different types of keys can co-exist at the
+/// same level and not interfere with each other, allowing for sorting in groups.
+/// Keys that evaluate as equal, or when compared with Widgets that don't have
+/// [Semantics], are given the default upper-start-to-lower-end geometric
```
Comment (goderbauer) [inline examples/catalog/lib/custom_a11y_traversal.dart]: I believe its currently top-left to bottom-right no matter of direction? 
https://github.com/flutter/engine/pull/4540/files#diff-be56efcdd49538955ac6ed238eda9ed2R447

File: examples/catalog/lib/custom_a11y_traversal.dart [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+/// ordering.
+///
+/// Since widgets are globally sorted by their sort key, the order does not have
+/// to conform to the widget hierarchy. Indeed, in this example, we traverse vertically
+/// first, but the widget hierarchy is a column of rows.
+///
+/// See also:
+///
+///  * [Semantics] for an object that annotates widgets with accessibility semantics.
```
Comment (goderbauer) [inline examples/catalog/lib/custom_a11y_traversal.dart]: maybe add: (including traversal order).

File: examples/catalog/lib/custom_a11y_traversal.dart [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+Sample Catalog
+
+Title: CustomTraversalExample
+
+Summary: An app that demonstrates a custom semantics traversal order.
+
+Description:
+This app presents a value selection interface where the fields can be
+incremented or decremented using spinner arrows. In accessibility mode, the
+widgets are traversed from one column to the next, starting with the column title, moving to
```
Comment (goderbauer) [inline examples/catalog/lib/custom_a11y_traversal.dart]: ... are traversed *in custom order* from ...

File: examples/catalog/lib/custom_a11y_traversal.dart [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        body: new Column(
+          mainAxisAlignment: MainAxisAlignment.center,
+          crossAxisAlignment: CrossAxisAlignment.center,
+          children: <Widget>[
+            new Semantics(
+              // Since this is the only sort key that the text has, it
+              // will be compared with the 'column' OrdinalSortKeys of all the
+              // fields, because the column sort keys are first in the other fields.
+              //
+              // an ordinal of "0.0" means that it will be traversed before column 1.
```
Comment (goderbauer) [inline examples/catalog/lib/custom_a11y_traversal.dart]: an -> An

File: examples/catalog/lib/custom_a11y_traversal.dart [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+              sortKey: const OrdinalSortKey(0.0),
+              child: const Text(
+                'How many pets do you own?',
+                textScaleFactor: 2.0,
```
Comment (goderbauer) [inline examples/catalog/lib/custom_a11y_traversal.dart]: Since all Text's (except for the AppBar one) set the textScaleFactor to 2, would it make sense to add a MediaQuery somewhere to the hierarchy that sets the textScaleFactor once?

File: examples/catalog/lib/custom_a11y_traversal.dart [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+              ),
+            ),
+            const Padding(padding: const EdgeInsets.symmetric(vertical: 10.0)),
+            new Row(
+              mainAxisAlignment: MainAxisAlignment.spaceAround,
+              children: <Widget>[
+                _makeFieldHeader(1, 0, Field.DOGS),
```
Comment (goderbauer) [inline examples/catalog/lib/custom_a11y_traversal.dart]: nit: I believe this would be easier to read if the int params are (required) named arguments. That way, I don't have to remember the order column and row params are specified in.

(here and below)

File: examples/catalog/lib/custom_a11y_traversal.dart [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+                _makeSpinnerButton(4, 2, Field.FISH, increment: false),
+              ],
+            ),
+            const Padding(padding: const EdgeInsets.symmetric(vertical: 10.0)),
+            new Semantics(
+              // Since this is the only sort key that the reset button has, it
+              // will be compared with the 'column' OrdinalSortKeys of all the
+              // fields, because the column sort keys are first in the other fields.
+              //
+              // an ordinal of "5.0" means that it will be traversed after column 4.
```
Comment (goderbauer) [inline examples/catalog/lib/custom_a11y_traversal.dart]: an -> An

File: examples/catalog/lib/custom_a11y_traversal.dart [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+enum Field { DOGS, CATS, FISH }
+
+String _fieldToName(Field field) {
+  switch (field) {
+    case Field.DOGS:
+      return 'Dogs';
+    case Field.CATS:
+      return 'Cats';
+    case Field.FISH:
+    default:
```
Comment (goderbauer) [inline examples/catalog/lib/custom_a11y_traversal.dart]: If you remove the "default", the analyzer would start complaining if we ever extend the enum and forgot to add the new field here.

File: examples/catalog/lib/custom_a11y_traversal.dart [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+    final String stringValue = '${_fieldToName(field)} $value';
+    final String increasedValue = '${_fieldToName(field)} ${value + 1}';
+    final String decreasedValue = '${_fieldToName(field)} ${value - 1}';
+
+    return new RowColTraversal(
+      rowOrder: rowOrder,
+      columnOrder: columnOrder,
+      child: new Center(
+        child: new Semantics(
+          onDecrease: onDecrease,
```
Comment (goderbauer) [inline examples/catalog/lib/custom_a11y_traversal.dart]: Oh nice! So in a11y mode you have two ways to manipulate the number: either use decrease/increase actions or click the up/down button!

**Review by goderbauer — APPROVED** [review]
LGTM

+1 for the extensive explanation of whats happening here.

File: examples/catalog/lib/custom_a11y_traversal.dart (hunk 3 above)
Comment (gspencergoog) [reply examples/catalog/lib/custom_a11y_traversal.dart]: Done.

File: examples/catalog/lib/custom_a11y_traversal.dart [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+/// first, but the widget hierarchy is a column of rows.
+///
+/// See also:
+///
+///  * [Semantics] for an object that annotates widgets with accessibility semantics
+///    (including traversal order).
+///  * [SemanticSortOrder] for the class that manages the sort order of a semantic node.
+///  * [SemanticSortKey] for the base class of all semantic sort keys.
+///  * [OrdinalSortKey] for a concrete sort key that sorts based on the given ordinal.
+class RowColTraversal extends StatelessWidget {
```
Comment (Hixie) [inline examples/catalog/lib/custom_a11y_traversal.dart]: drive-by nit: Prefer avoiding abbreviations, so this would be RowColumnTraversal
